### DISCUSSION TITLE
Valid versioning syntax when specifying Python version in tox.ini

### DISCUSSION THREAD

**Discussion by inwaves** [ymyzk/tox-gh-actions#153] [Q&A] [answered]
**Question**
Hi! Does tox-gh-actions support `x.y-dev` syntax when specifying the Python version in `tox.ini`?

**Context**
For context, this is a downstream issue in [PyCQA/flake8-bugbear](https://github.com/PyCQA/flake8-bugbear/issues/300), where in our config we're able to specify a release candidate version of 3.11 (using the syntax detailed [here](https://github.com/actions/setup-python/blob/main/docs/advanced-usage.md#using-the-python-version-input)), for GH Actions and install it, but this isn't being picked up by tox as matching any environments from the config envlist.

For an example of the failure, see [this CI run](https://github.com/inwaves/flake8-bugbear/actions/runs/3635937491/jobs/6135474435#step:6:12). Our config is [here](https://github.com/PyCQA/flake8-bugbear/blob/71410094e3c34b23e8c12c081b9a7901c2cc6ffe/tox.ini#L12).

**Additional context**
I've been able to run this by replacing `3.11.0-rc - 3.11: py311` with `3.11: py311`, but in the future we may want to support Python versions before full release. Looking at the source code [here](https://github.com/ymyzk/tox-gh-actions/blob/abadecd4b1e330bb6fa554e625c87cdd07a0634e/src/tox_gh_actions/plugin.py#L231), I'm thinking that tox-gh-actions is comparing the Python version requested by the config with the `major.minor`-formatted version of the actual distribution. To support this, it might be necessary to take parse something like `major.minor.micro-releaselevel` (as supplied by `sys.versioninfo`).

**Comment by ymyzk** [comment, marked as answer]
Hello @inwaves, thanks for filing a ticket.

I'd like to understand your requirement a bit more carefully.

[As explained in our document](https://github.com/ymyzk/tox-gh-actions/tree/abadecd4b1e330bb6fa554e625c87cdd07a0634e#factor-conditional-settings-python-version), tox-gh-actions can be used with pre-released Python versions. The point is that we shouldn't specify micro and release level in the tox-gh-actions' config even when we're using a pre-released version. tox-gh-actions doesn't care whether a Python version is pre-release or not at this point.

`.github/workflows/<workflow>.yml`:

```
...
jobs:
  build:
    strategy:
      matrix:
        python-version: ['3.10', 3.11.0-beta.3]
...
```

`tox.ini`:
```
[tox]
envlist = py310, py311

[gh-actions]
python =
    3.10: py310
    3.11: py311
    # The following won't work
    # 3.11-beta.3: py311
    # 3.11-dev: py311

[testenv]
...
```

Do you have a requirement to specify micro and release level versions precisely in the tox-gh-actions' config like the following?

`tox.ini`:
```
[tox]
envlist = py310, py311b2, py311b3

[gh-actions]
python =
    3.10: py310
    3.11-beta.2: py311b2
    3.11-beta.3: py311b3
[testenv]
...
```

**Comment by inwaves** [comment]
Hi! I think the idea for specifying the entire version is that we can use pre-release versions like 3.11.0-rc. Are you saying that I can specify the pre-release Python version in my workflow YAML and point to it in `tox.ini` using the shorthand version? (This is how I'm reading your examples above.) If this is the case, then we should be able to just do that – no preference for specifying the entire version in `tox.ini`.

**Comment by ymyzk** [comment]
> Are you saying that I can specify the pre-release Python version in my workflow YAML and point to it in tox.ini using the shorthand version? (This is how I'm reading your examples above.)

That's right. We can (rather must) use the shorthand version in `tox.ini`.

**Comment by inwaves** [comment]
Thanks for getting back to me, I'm happy to close this.
